mds: clear MDCache::rejoin_*_q queues before recovering file inodes

lxbsz opened the pull request:
If the monitor sends rejoin mdsmap twice just before the first time
hasn't finished yet, it may run identify_files_to_recover() twice.

Since the rejoin_recover_q and rejoin_check_q are vector so there
could be duplicated inodes.

Fixes: https://tracker.ceph.com/issues/53741
Signed-off-by: Xiubo Li <xiubli@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

joscollin reviewed (commented):
This has related errors in make check.

lxbsz commented:
> This has related errors in make check.

Fixed it, thanks @joscollin

joscollin commented on src/mds/MDCache.h:
```diff
@@ -1206,7 +1206,7 @@ class MDCache {
   std::map<mds_rank_t, std::set<CInode*> > rejoin_unlinked_inodes;
 
-  std::vector<CInode*> rejoin_recover_q, rejoin_check_q;
+  std::set<CInode*> rejoin_recover_q, rejoin_check_q;
```
How about using a flat_set instead?

lxbsz replied:
Since the MDS won't do random access to these two queues, IMO it doesn't make much sense to use `flat_set` (?).

joscollin replied:
Right. There is iteration, but called only once.

ukernel commented:
if identify_files_to_recover can be called multiple times, why not clear these two queue at very beginning of identify_files_to_recover()

lxbsz commented:
> if identify_files_to_recover can be called multiple times, why not clear these two queue at very beginning of identify_files_to_recover()

Sounds good and fixed it.

ukernel commented:
> > if identify_files_to_recover can be called multiple times, why not clear these two queue at very beginning of identify_files_to_recover()
> 
> Sounds good and fixed it.

no longer need to be set

lxbsz commented:
> > > if identify_files_to_recover can be called multiple times, why not clear these two queue at very beginning of identify_files_to_recover()
> > 
> > 
> > Sounds good and fixed it.
> 
> no longer need to be set

Removed that change. Thanks.

vshankar reviewed (approved):
Looks good.

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Jun-13
